File: toolbox_starter_solution/week3_greedy_algorithms/3_car_fueling/car_fueling.py
from sys import stdin


def min_refills(distance, tank, stops):
    stops.insert(0, 0)  # Insert the starting point at the beginning of the stops list.
    stops.append(distance)  # Add the destination as the final stop.
    
    num_refills = 0  # Initialize the number of refills to 0.
    current_refill = 0  # Initialize the index of the current gas station.
    
    while current_refill < len(stops) - 1:
        last_refill = current_refill
        
        # Find the farthest gas station the car can reach from the current one.
        while current_refill < len(stops) - 1 and stops[current_refill + 1] - stops[last_refill] <= tank:
            current_refill += 1
        
        if current_refill == last_refill:
            return -1  # Unable to reach the next gas station.
        
        if current_refill < len(stops) - 1:
            num_refills += 1
    
    return num_refills

if __name__ == '__main__':
    d, m, _, *stops = map(int, stdin.read().split())
    # d, m, stops = 950, 400, [200, 375, 550, 750]
    print(min_refills(d, m, stops))

File: toolbox_starter_solution/week3_greedy_algorithms/3_car_fueling/car_fueling.cpp
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::max;
using std::vector;

int compute_min_refills(int dist, int tank, vector<int> &stops)
{
    // write your code here
    if (dist <= tank)
    {
        return -1;
    }
    int numRefills = 0;
    int currentRefill = 0;
    int lastRefill = 0;
    int n = stops.size();
    int cur_tank = tank - stops[0];
    while (currentRefill < n - 1)
    {
        // lastRefill = currentRefill;
        while (currentRefill < n - 1 && stops[currentRefill + 1] - stops[currentRefill] <= cur_tank)
        {
            cur_tank = cur_tank - (stops[currentRefill+1] - stops[currentRefill]);
            currentRefill++;
        }

        if (currentRefill < n - 1)
        {
            numRefills++;
            cur_tank = tank;
        }

        if (stops[currentRefill] + cur_tank < stops[currentRefill + 1])
        {
            return -1;
        }
    }
    if (stops[currentRefill] + cur_tank >= dist)
    {
        return numRefills;
    }
    else if (stops[currentRefill] + tank >= dist)
    {
        return numRefills + 1;
    }
    else
    {
        return -1;
    }
}

int main()
{
    int d = 0;
    cin >> d;
    int m = 0;
    cin >> m;
    int n = 0;
    cin >> n;

    vector<int> stops(n);
    for (size_t i = 0; i < n; ++i)
        cin >> stops.at(i);

    cout << compute_min_refills(d, m, stops) << "\n";

    return 0;
}
